Comp Threads and JIT Server options post restore

dsouzai opened the pull request:
* Add the concept of Number of Allocated Threads
    * In CRIU mode, the number of threads allocated is `TR::CompilationInfo::MAX_CLIENT_USABLE_COMP_THREADS` (unless it exceeds the number of code caches).
    * In non-CRIU mode, the number of threads allocated is the same as the number of usable threads.
* Refactor JITServer options processing code so it can be reused post-restore
* Process JITServer options post restore
* Enable the post restore options processing code

dsouzai commented on runtime/compiler/control/OptionsPostRestore.cpp:
```diff
@@ -331,79 +253,45 @@ J9::OptionsPostRestore::iterateOverExternalOptions()
 J9::OptionsPostRestore::processJitServerOptions()
    {
+#if defined(J9VM_OPT_JITSERVER)
    if (_argIndexUseJITServer >= _argIndexDisableUseJITServer)
```
It is worth noting that this may need to be changed in the future to account for the fact that in CRIU mode, JITServer is only enabled by default in non-portable mode (https://github.com/eclipse-openj9/openj9/pull/15674)

dsouzai commented:
@mpirvu could you please review?

mpirvu commented on runtime/compiler/control/OptionsPostRestore.cpp:
```diff
@@ -150,7 +154,19 @@ J9::OptionsPostRestore::iterateOverExternalOptions()
             {
-            // Need to allocate more if needed
+            char *compThreadsOption = J9::Options::_externalOptionStrings[J9::ExternalOptions::XcompilationThreads];
+            int32_t argIndex = FIND_AND_CONSUME_RESTORE_ARG(EXACT_MEMORY_MATCH, compThreadsOption, 0);
```
This doesn't sound right. `EXACT_MEMORY_MATCH` should be used for memory values.

dsouzai replied:
I just copied this code from [preProcessCompilationThreads](https://github.com/eclipse-openj9/openj9/blob/9bf20a075d01f0f2377a15c0adad686aa648259c/runtime/compiler/control/J9Options.cpp#L2077-L2078), so it seems we're already ok with `EXACT_MEMORY_MATCH`.

mpirvu commented on runtime/compiler/control/OptionsPostRestore.cpp:
```diff
@@ -150,7 +154,19 @@ J9::OptionsPostRestore::iterateOverExternalOptions()
+
+               if (ret == OPTION_OK && numCompThreads > 0)
+                  {
+                  TR::Options::_numUsableCompilationThreads = numCompThreads;
```
Should we check that it does not exceed the number of allocated threads before checkpoint?

mpirvu commented on runtime/compiler/control/OptionsPostRestore.cpp:
```diff
@@ -154,12 +163,11 @@ J9::OptionsPostRestore::iterateOverExternalOptions()
             {
-            char *compThreadsOption = J9::Options::_externalOptionStrings[J9::ExternalOptions::XcompilationThreads];
-            int32_t argIndex = FIND_AND_CONSUME_RESTORE_ARG(EXACT_MEMORY_MATCH, compThreadsOption, 0);
+            int32_t argIndex = FIND_AND_CONSUME_RESTORE_ARG(EXACT_MEMORY_MATCH, optString, 0);
```
`EXACT_MEMORY_MATCH` does not look right here

dsouzai replied:
Same response as https://github.com/eclipse-openj9/openj9/pull/16769#discussion_r1120255959

mpirvu reviewed (commented):
I have some inline comments

dsouzai commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -3119,7 +3159,11 @@ TR::CompilationInfo::startCompilationThread(int32_t priority, int32_t threadId,
       {
       getCompilationMonitor()->enter();
-      _numCompThreads++;
+      if (_numCompThreads < TR::Options::_numUsableCompilationThreads)
```
I don't think it would fail. However, the reason for this `if` is so that we don't increment `_numCompThreads` past `_numUsableCompilationThreads`; we need to run `startCompilationThread` for all allocated threads, but we also need to maintain the distinction between the `_numCompThreads` and `_numAllocatedCompThreads`.

dsouzai commented:
@mpirvu good for review again (see [force push](https://github.com/eclipse-openj9/openj9/pull/16769)). I updated `setNumUsableCompilationThreadsPostRestore` to take in the parameter by reference so that it updates `TR::Options::_numUsableCompilationThreads`; this ensures that we do the check you mentioned in https://github.com/eclipse-openj9/openj9/pull/16769#discussion_r1119388807.

Also, I added the code to actually run this code post restore so that we start exercising this stuff.

mpirvu commented on runtime/compiler/control/J9Options.cpp:
```diff
@@ -2712,6 +2714,15 @@ J9::Options::fePostProcessJIT(void * base)
+   if (!javaVM->internalVMFunctions->isJITServerEnabled(javaVM))
+#endif
+      {
+      _numAllocatedCompilationThreads = TR::CompilationInfo::MAX_CLIENT_USABLE_COMP_THREADS;
```
This is server mode, right? In that case shouldn't this be `MAX_SERVER_USABLE_COMP_THREADS`

dsouzai replied:
No, this is for `!javaVM->internalVMFunctions->isJITServerEnabled(javaVM)` so it's only for the client; unless i'm misunderstanding what `isJITServerEnabled` is supposed to be checking for?

mpirvu replied:
Right, I missed the `!`

mpirvu commented on runtime/compiler/control/OptionsPostRestore.cpp:
```diff
@@ -331,79 +250,45 @@ J9::OptionsPostRestore::iterateOverExternalOptions()
-            }
+         char *address = NULL;
+         GET_OPTION_VALUE_RESTORE_ARGS(_argIndexJITServerAddress, '=', &address);
+         _compInfo->getPersistentInfo()->setJITServerAddress(address);
```
This is arguably the most important option for a JITServer client. Could you please test locally and make sure it works as designed?

mpirvu commented on runtime/compiler/runtime/JITServerStatisticsThread.cpp:
```diff
@@ -115,12 +115,13 @@ static int32_t J9THREAD_PROC statisticsThreadProc(void * entryarg)
             TR_VerboseLog::writeLine(TR_Vlog_JITServer, "Compilation Queue Size: %d", compInfo->getMethodQueueSize());
             TR_VerboseLog::writeLine(TR_Vlog_JITServer, "Number of clients : %u", compInfo->getClientSessionHT()->size());
             TR_VerboseLog::writeLine(TR_Vlog_JITServer, "Total compilation threads : %d", compInfo->getNumUsableCompilationThreads());
+            TR_VerboseLog::writeLine(TR_Vlog_JITServer, "Total allocated compilation threads : %d", compInfo->getNumAllocatedCompilationThreads());
```
There is no value in this line since in server mode we do `TR::Options::_numAllocatedCompilationThreads = TR::Options::_numUsableCompilationThreads;` You can delete it.

mpirvu reviewed (commented):
I have a few more suggestions. Please see inline.

dsouzai commented:
@mpirvu addressed review comments (see [force push](https://github.com/eclipse-openj9/openj9/compare/13310a921884db06c3eea53a14ca5cab4271b719..60bfe7b0b7ba9467425db6b26f1f14fc0bdf4b91)).

Also I ran some tests:


Specifying nothing (defaults to localhost) or the machine name:
```
 (warm) Compiling HelloInstantOn.jitMe()V  OrdinaryMethod j9m=0000000000191700 remote t=15206 compThreadID=3 memLimit=1572864 KB freePhysicalMemory=1534 MB
#JITServer: Client sending compReq seqNo=1 to server for method HelloInstantOn.jitMe()V @ warm.
#JITServer: t= 15206 Connected to a server (serverUID=17563505271461179976)
#JITServer: Client successfully loaded method HelloInstantOn.jitMe()V @ warm following compilation request. [metaData=00007F3848A86038, startPC=00007F3D812B4058]
+ (warm) HelloInstantOn.jitMe()V @ 00007F3D812B4058-00007F3D812B48E8 OrdinaryMethod - Q_SZ=3 Q_SZI=3 QW=26 j9m=0000000000191700 bcsz=23 OSR remote time=86227us mem=[region=448 system=2048]KB compThreadID=3 CpuLoad=0%(0%avg) JvmCpu=55% queueTime=86978us
```

Specifying `-XX:JITServerAddress=badaddr`:
```
 (warm) Compiling HelloInstantOn.jitMe()V  OrdinaryMethod j9m=0000000000191700 remote t=9436 compThreadID=4 memLimit=1572864 KB freePhysicalMemory=1523 MB
#FAILURE:  JITServer::StreamFailure: Cannot resolve server name: Name or service not known for HelloInstantOn.jitMe()V @ warm
#JITServer: t=  9436 Could not connect to a server. Next attempt in 2000 ms.
#JITServer: t=  9436 Resetting activation policy to AGGRESSIVE because client has lost connection to server
#JITServer: compThreadID=4 JITServer StreamFailure: Generic stream failure
! (warm) HelloInstantOn.jitMe()V Q_SZ=3 Q_SZI=3 QW=26 j9m=0000000000191700 time=29751us compilationStreamFailure memLimit=1572864 KB freePhysicalMemory=1523 MB mem=[region=64 system=2048]KB compThreadID=4
 (warm) Compiling HelloInstantOn.jitMe()V  OrdinaryMethod j9m=0000000000191700  t=9436 compThreadID=4 memLimit=1572864 KB freePhysicalMemory=1523 MB
+ (warm) HelloInstantOn.jitMe()V @ 00007FEF322B4058-00007FEF322B4801 OrdinaryMethod - Q_SZ=3 Q_SZI=3 QW=26 j9m=0000000000191700 bcsz=23 OSR time=20324us mem=[region=3968 system=4096]KB compThreadID=4 CpuLoad=0%(0%avg) JvmCpu=76% queueTime=50911us
```

mpirvu commented:
jenkins test sanity all jdk17
